main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer1.scala

```diff
@@ -0,0 +1,580 @@
+        (OccurrenceAst1.Expr.Match(e, rs, tpe, eff, loc), increment(o3))
+
+      case MonoAst.Expr.VectorLit(exps, tpe, eff, loc) =>
+        val (es, o) = visitExps(exps)
+        (OccurrenceAst1.Expr.VectorLit(es, tpe, eff, loc), increment(o))
+
+      case MonoAst.Expr.VectorLoad(exp1, exp2, tpe, eff, loc) =>
+        val (e1, o1) = visit(exp1)
+        val (e2, o2) = visit(exp2)
+        val o3 = combineInfo(o1, o2)
+        (OccurrenceAst1.Expr.VectorLoad(e1, e2, tpe, eff, loc), increment(o3))
+
+      case MonoAst.Expr.VectorLength(exp, loc) =>
+        val (e, o) = visit(exp)
+        (OccurrenceAst1.Expr.VectorLength(e, loc), increment(o))
+
+      case MonoAst.Expr.Ascribe(exp, tpe, eff, loc) =>
+        val (e, o) = visit(exp)
+        (OccurrenceAst1.Expr.Ascribe(e, tpe, eff, loc), increment(o))
+
+      case MonoAst.Expr.Cast(exp, declaredType, declaredEff, tpe, eff, loc) =>
+        val (e, o) = visit(exp)
+        val o1 = o :+ sym0 -> Dangerous
+        (OccurrenceAst1.Expr.Cast(e, declaredType, declaredEff, tpe, eff, loc), increment(o1))
+
+      case MonoAst.Expr.TryCatch(exp, rules, tpe, eff, loc) =>
+        val (e, o1) = visit(exp)
+        val (rs, o2) = visitTryCatchRules(rules)
+        val o3 = combineInfo(o1, o2) :+ sym0 -> DontInline
+        (OccurrenceAst1.Expr.TryCatch(e, rs, tpe, eff, loc), increment(o3))
+
+      case MonoAst.Expr.TryWith(exp, effUse, rules, tpe, eff, loc) =>
+        val (e, o1) = visit(exp)
+        val (rs, o2) = visitTryWithRules(rules)
+        val o3 = combineInfo(o1, o2) :+ sym0 -> DontInline
+        (OccurrenceAst1.Expr.TryWith(e, effUse, rs, tpe, eff, loc), increment(o3))
+
+      case MonoAst.Expr.Do(op, exps, tpe, eff, loc) =>
+        val (es, o) = visitExps(exps)
+        (OccurrenceAst1.Expr.Do(op, es, tpe, eff, loc), increment(o))
+
+      case MonoAst.Expr.NewObject(name, clazz, tpe, eff, methods, loc) =>
+        val (ms, o) = visitJvmMethods(methods)
+        (OccurrenceAst1.Expr.NewObject(name, clazz, tpe, eff, ms, loc), increment(o))
+
+    }
+
+    /**
+      * Performs occurrence analysis on a list of expressions `exps` and merges occurrences.
+      * Captures `sym0`.
+      */
+    def visitExps(exps: List[MonoAst.Expr])(implicit letBinding: Option[VarSym]): (List[OccurrenceAst1.Expr], OccurInfo) = {
+      val (es, o1) = exps.map(visit).unzip
+      val o2 = o1.foldLeft(OccurInfo.Empty)(combineInfo)
+      (es, o2)
+    }
+
+    def combineApplyCloInfo(occurInfo0: OccurInfo, exp0: OccurrenceAst1.Expr): OccurInfo = exp0 match {
+      case OccurrenceAst1.Expr.ApplyAtomic(AtomicOp.Closure(sym), _, _, _, _) =>
+        occurInfo0 :+ sym -> Once
+      case _ => occurInfo0
+    }
+
+    def visitAtomicOp(op0: AtomicOp, occurInfo0: OccurInfo)(implicit letBinding: Option[VarSym]): OccurInfo = op0 match {
+      case AtomicOp.Is(sym) if sym.name == "Choice" =>
+        occurInfo0 :+ sym0 -> DontInline
+
+      case AtomicOp.HoleError(_) => letBinding match {
```

Comment:
im not a fan of this but i can see why you're kind of forced into it by the current stdlib / language. hmmm